[Examples/Docs] Tic-tac-toe (#18558)
MystenLabs/sui: examples/tic-tac-toe/README.md

# Tic tac toe

This is an end-to-end example for on-chain tic-tac-toe. It includes:

- A [Move package](./move), containing two protocols for running a game of
  tic-tac-toe. One that uses shared objects and consensus and another

>   that uses owned objects, and the fast path.

Reviewer: Maybe define "fast path" in a couple of words?

- A [React front-end](./ui), in TypeScript built on top of
  `create-react-dapp`, using the TS SDK and `dapp-kit`.
- A [Rust CLI](./cli), using the Rust SDK.
- [Scripts](./scripts) to publish packages and update configs used
  while building the front-end and CLI.

## Shared tic tac toe

In the shared protocol, player X creates the `Game` as a shared object

> and players take it in turns to place marks. Once the final move is

Reviewer: ```suggestion
and players take turns to place marks. Once the final move is
```

receives the `Mark` on the game and places it.

Control of who makes the next move is decided using a `TurnCap`.
Initially Player X has the `TurnCap`. This capability must be consumed
to create a `Mark`, and when the admin places the mark, a new
`TurnCap` is created and sent to the next player, if the game has not
ended yet.

As in the shared protocol, once the game has ended, a `Trophy` is sent
to any winning player. Unlike the shared protocol, only the admin can
clean-up the Game once it has finished, because only they have access
to it.

``` mermaid
sequenceDiagram
    activate Player X
    Player X->>Admin: new: Game
    Player X->>Player X: [TurnCap]
    deactivate Player X
    Player X->>Game: send_mark: Mark
    activate Admin
    Admin->>Game: place_mark
    Admin->>Player O: [TurnCap]
    deactivate Admin
    Player O->>Game: send_mark: Mark
    activate Admin
    Admin->>Game: place_mark
    Admin->>Player X: [TurnCap]
    deactivate Admin
    Player X->>Game: ...
    Admin->>Game: ...
    Player O->>Game: ...
    activate Admin
    Admin->>Game: place_mark
    Admin->>Player O: [Trophy]
    deactivate Admin
    Admin->>Game: burn
```

> ## Multi-sig tic-tac-toe

Reviewer: ```suggestion
## Multisig tic-tac-toe
```
It's multisig in the doc above (and in docs in general). It's consistent here so tomato, tomato.

can choose not to place a move, or simply forget to). That third party
may also need to run a service that keeps track of marks sent to games
in order to apply them promptly, which adds complexity.

There is an alternative approach, which leverages Sui's support for
**multi-sigs** and **sponsored transactions**. Instead of entrusting
the Game to a third party, it is sent to an address owned by a 1-of-2
multisig, signed for by Player X and Player O.

Play proceeds as in the owned protocol, except that the Admin is the
multi-sig account. On each turn, the current player runs a transaction
as themselves to send the mark, and then runs a transaction on behalf
of the multi-sig to place it.

Once play has finished, either player can run a transaction on behalf
of the multi-sig account to `burn` the game. As the player is the
sponsor, they will receive the storage rebate for performing the
clean-up.

The multi-sig account does not own anything other than the game object
(it does not have any gas coins of its own), so the player sponsors
the transaction, using one of its own gas coins.

Sharing a resource while avoiding consensus by transfering it to a

> multi-sig account can be generalised from two accounts to up to ten

Reviewer: ```suggestion
multi-sig account can be generalized from two accounts to a max of ten
```
"to up to" just reads odd to me. Also, 'merica!

(the limit being the number of keys that can be associated with one
multi-sig).

In order to create a multi-sig, the public keys of all the signers
needs to be known. Each account address on Sui is the hash of a public
key, but this operation cannot be reversed, so in order to start a
multi-sig game, players must exchange public keys instead of
addresses.

## Pros and cons

The shared protocol's main benefit is that its on-chain logic and
client integration are straightforward, and its main downside is that
it relies on consensus for ordering.

In contrast, the owned protocol uses only fast-path transactions (no

> consensus), but its on-chain logic is more complicated because it

Reviewer: Maybe use this parenthetical in the first note (no consensus) and then not needed here.

needs to manage the `TurnCap`, and its off-chain logic is complicated
either by a third party service to act as an Admin, or a multi-sig and

> sponsored transaction set-up. When using multi-sig, care also needs to

Reviewer: ```suggestion
sponsored transaction setup. When using multi-sig, care also needs to
```
nit
